File: medium/3446.java
import java.util.*;

class Solution {
    public int[][] sortMatrix(int[][] grid) {

        // bottom left - non decreasing
        int n = grid.length;
        for (int row = 0; row < n; row++)
            solve(grid, row, 0, n, true);

        // top right -- non increasing
        for (int col = 1; col < n; col++)
            solve(grid, 0, col, n, false);

        return grid;

    }

    private void solve(int[][] grid, int row, int col, int n, boolean desc) {

        List<Integer> list = new ArrayList<>();

        int i = row;
        int j = col;

        while (i < n && j < n) {
            list.add(grid[i][j]);
            i++;
            j++;
        }

        if (desc)
            Collections.sort(list, Collections.reverseOrder());
        else
            Collections.sort(list);

        i = row;
        j = col;

        for (int val : list) {
            grid[i][j] = val;
            i++;
            j++;
        }

    }
}

File: medium/3446.py
class Solution:
    def sortMatrix(self, grid: List[List[int]]) -> List[List[int]]:
       n = len(grid)

       # bottom left - non decreasing
       for row in range(n):
           self.solve(grid,row,0,n,True)

       # top right - non increasing
       for col in range(1,n):
           self.solve(grid,0,col,n,False)

       return grid


    def solve(self, grid: List[List[int]], row: int, col: int, n: int, desc: bool) -> None:
        vals = []
        i,j=row,col

        while i<n and j<n:
            vals.append(grid[i][j])
            i+=1
            j+=1

        if desc:
            vals.sort(reverse=True)
        else:
            vals.sort()
        i,j=row,col

        for val in vals:
            grid[i][j]=val
            i+=1
            j+=1


